### PR TITLE
EEP: Trait Observers

### PR DESCRIPTION
**Pull request by corranwebster**
This is a proposal to create an alternative observation pattern to allow migration away from trait listeners and the issues that they have without breaking existing code.

### CODE REVIEW

**Comment by prabhuramachandran** [comment]
Thank you for this.  I have a feature request which may not fit here but here goes anyway. For recording trait changes, I would find it very useful to set a priority of when the handler gets called.  For example, I would like to add an observer that gets called first before any others are called.  This would clean up the recorded data.

File: docs/source/eeps/eep-3.rst [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+* ``i.+foo`` -> ``obs('i', obs('', metadata={'foo': not_none}))``
+* ``[x,y].z`` -> ``obs(['x', 'y'], 'z')``
+* ``[x,y.z]`` -> ``obs(['x', obs('y', 'z')])``
+
+
+Implementation
+==============
+
+Much if this can be implemented using the existing notification system.  At
+the core, cTrait classes have a list of "notifiers" which are callables that
```
Comment (mdickinson) [inline docs/source/eeps/eep-3.rst]: nitpick: replace "classes" with "instances" here?

**Comment by mdickinson** [comment]
@prabhuramachandran I think this is somewhat orthogonal to this particular proposal. Are you interested in fleshing out the idea in a new EEP?

**Comment by corranwebster** [comment]
@prabhuramachandran That could absolutely fit here.

Do you have thoughts on what the API and implementation might look like?

One possibility that springs to mind is an optional numeric `priority` parameter, which is used to keep the notifiers list sorted (with all non-priority items sorted to the end)?  But it may not interact well with existing notifiers, so you'd need to switch to the new system to get the full benefit.

**Comment by corranwebster** [comment]
Summarizing an in-person conversation with @mdickinson: priorities could be done without going making changes to the ctraits.c level (which was Mark's main concern), but you would have to allow for the possibility that someone else might manually change the ordering of notifiers; for going first absolutely (ie. some sort of "pre-notifiy" or "post-getattr" hook), that would require changes at the ctraits level, and probably should be its own EEP.

**Comment by mdickinson** [comment]
Merging. If we want to expand this EEP to cover priorities, that can easily be done in another PR.

**Pull request merged**